### PR TITLE
Linkouts

### PR DESCRIPTION
**Pull request by adf-ncgr**
adding a new microservice for linkouts (including currently linkouts for gene and genomic_region); seemed to make sense to handle everything with a single implementation exposing multiple endpoints for the different types of input, since the plan would be to store all configuration in the same datastore READMEs.

### CODE REVIEW

**Review by alancleary — CHANGES_REQUESTED** [review]
Thanks for the PR, @adf-ncgr. I have quite a few comments. Perhaps it's worth us doing a code review together?

**Comment by adf-ncgr** [comment]
@alancleary sounds good, just let me know when

**Review by alancleary — CHANGES_REQUESTED** [review]
Please make the changes discussed during code review + add automated build and deployment.

File: linkouts/linkouts/__main__.py [hunk 1]
```diff
@@ -40,13 +49,37 @@ def parseArgs():
     version=f'%(prog)s {linkouts.__version__}',
   )
 
+  # logging args
+  loglevel_envvar = 'LOG_LEVEL'
```
Comment (alancleary) [inline linkouts/linkouts/__main__.py:53]: The log level and log file info isn't actually used to configure the Python logging module!

File: linkouts/README.md [hunk 2]
```diff
@@ -1,19 +1,49 @@
...
+  -
+    method: GET
+    href: https://vigna.legumeinfo.org/tools/jbrowse2?config=vigna.json&session=spec-%7B%22views%22%3A%5B%7B%22assembly%22%3A%22vigun.CB5-2.gnm1%22%2C%22loc%22%3A%22{GENOMIC_REGION_CHR_ID}%3A{GENOMIC_REGION_START}-{GENOMIC_REGION_END}%22%2C%22type%22%3A%20%22LGV%22%2C%22tracks%22%3A%5B%22gene_models_main%22%5D%7D%5D%7D
+    text: "View {GENOMIC_REGION} in JBrowse2"
+```
 
 ## Use
 
-The microservice can be queried via HTTP POST or gRPC.
+The microservice can be queried via HTTP GET.
```
Comment (alancleary) [inline linkouts/README.md:45]: The HTTP server still has `POST` endpoints. Is there a reason they're not mentioned in the readme?

File: linkouts/README.md [hunk 3]
```diff
@@ -1,19 +1,49 @@
...
 
 The default request URLs are :
-- `localhost:8080/gene_linkouts`
-- `localhost:8080/genomic_region_linkouts`
+- `localhost:8080/gene_linkouts?genes=<comma-delimited geneid list>`
+- `localhost:8080/genomic_region_linkouts=<comma-delimited list of seqid:start-end`
```
[the remark this answers is not in the document]
Comment (adf-ncgr) [reply linkouts/README.md:49]: Huh, for some reason I do not seem to get emails from github regarding these comments (though I do get every other notification from github that I can imagine). So, just seeing all these review comments now and will try to address.

[the remark this answers is not in the document]
Comment (alancleary) [reply linkouts/README.md:49]: Weird. I'll be sure to mention you in future comments.

[the remark this answers is not in the document]
Comment (adf-ncgr) [reply linkouts/README.md:49]: Probably no need, I think I figured out the notifications settings needed- not sure why they treat PR reviews as a separate category, but I didn't previously have it set, now I do:
image

File: linkouts/README.md (hunk 2 above)
Comment (adf-ncgr) [reply linkouts/README.md:45]: The reason is that I haven't really implemented it, mostly meaning that there is no specification for how to form POST body content in the LINKOUTS.*.yml.

File: linkouts/linkouts/__main__.py [hunk 4]
```diff
@@ -55,25 +88,27 @@ def parseArgs():
...
+  loop = uvloop.new_event_loop()
+  asyncio.set_event_loop(loop)
 
   # run the program
-  #loop.create_task(main_coroutine(args))
   main_coroutine(args)
-  #loop.run_forever()
+
+# the asyncio exception handler that will initiate a shutdown
+def handleException(loop, context):
```
[the remark this answers is not in the document]
Comment (adf-ncgr) [reply linkouts/linkouts/__main__.py:107]: good catches on this and the above- my monkey-see-do skills seem to be sorely lacking!

**Comment by adf-ncgr** [comment]
@alancleary a couple of questions regarding your comment:
- re: the env variables in the compose file, I did document DATA in the "Running" section of the README, and PORT has a default specified; can you elaborate on what else you think is needed here?
- re: the addition of async to methods, I am probably not understanding very well but I think I'd added them to get around this error:`TypeError: object list can't be used in 'await' expression`; is there a better way of dealing with this?

happy to discuss this as part of the shimming follow-up if you prefer.

File: linkouts/linkouts/__main__.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+          f'the {logfile_envvar} environment variable).'))
+
+
+  # Async HTTP args
+  hhost_envvar = 'HTTP_HOST'
+  parser.add_argument('--hhost', action=EnvArg, envvar=hhost_envvar, type=str, default='127.0.0.1', help=f'The HTTP server host (can also be specified using the {hhost_envvar} environment variable).')
+  hport_envvar = 'HTTP_PORT'
+  parser.add_argument('--hport', action=EnvArg, envvar=hport_envvar, type=str, default='8880', help=f'The HTTP server port (can also be specified using the {hport_envvar} environment variable).')
+  lglob_root_envvar = ''
+  parser.add_argument('--lglob_root', action=EnvArg, envvar=lglob_root_envvar, type=str, default='/data', help=f'The root folder to be searched for linkouts.*.yml files containing linkout specifications')
```
Comment (alancleary) [inline linkouts/linkouts/__main__.py:84]: This should mention the lglob environment variable.

File: linkouts/linkouts/__main__.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    'format': '%(asctime)s,%(msecs)d %(levelname)s: %(message)s',
+    'datefmt': '%H:%M:%S',
+    'level': LOG_LEVELS[args.log_level],
+  }
+  if 'log_file' in args:
+    log_config['filename'] = args.log_file
+  logging.basicConfig(**log_config)
+
+  # initialize asyncio
+  #uvloop.install()
```
Comment (alancleary) [inline linkouts/linkouts/__main__.py:111]: This line should be removed; that uvloop function is deprecated so it won't even work in future releases.

File: linkouts/README.md (hunk 2 above)
Comment (alancleary) [reply linkouts/README.md:45]: Maybe add another `data` field that's required if the `method` is `POST`? The value of the `data` field could just be an object describing the payload.

I don't consider this a blocker so I'll mark this conversation as resolved.

**Comment by alancleary** [comment]
> re: the env variables in the compose file, I did document DATA in the "Running" section of the README, and PORT has a default specified; can you elaborate on what else you think is needed here?

After revisiting the code I only spotted one place in the argparse help strings where this was an issue. I commented on the specific line. That's pretty minor so I must've been mistaken before.

> re: the addition of async to methods, I am probably not understanding very well but I think I'd added them to get around this error: `TypeError: object list can't be used in 'await' expression`; is there a better way of dealing with this?

When you define a function/method as `async` then you need to use the `await` keyword when calling it. The error means you got rid of the `async` keyword on the `process_genes` method definition but kept the `await` keyword on the method call in `http_server.py`. It's not the end of the word if you leave the methods as `async` but it's misleading because they're not doing anything asynchronously, i.e. there's no advantage to them being defined as `async`.

**Comment by adf-ncgr** [comment]
OK, thanks for the clarification, I had copied the async/await by analogy with the search microservice, without thinking about the salient difference that the latter has the async interaction with redis. I think I've now addressed your last comments. With  respect to POST and adding a data field, yes I am leaning towards basically making the fields in the linkout spec mirror what the fetch API allows (thus potentially including things like headers for form-encoding and whatnot). But, I am also wondering if the specification will need some heavier templating magic like support for iteration over a set of values in the request and mapping it into a new structure for the POST body. Going to ponder a bit more before tackling it.

**Review by alancleary — APPROVED** [review]
Looks go to me. Feel free to merge if you've already tested, etc.

**Pull request merged**